src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+    flex_nd_vec bitmap, Point start, Point end, float stroke_width) {
+    swap_points(&start, &end);
+    float slope = (end.y - start.y)/(end.x - start.x);
+    int x1 = (int)(start.x);
+    int y1 = (int)(start.y);
+    int x2 = (int)(end.x);
+    int y2 = (int)(end.y);
+    int pad = (int)(stroke_width/2);
+    size_t dimension = bitmap.shape()[1];
+    if (x1 == x2) {
+        // just paint the y axis
+        int min_y = MIN(y1, y2);
+        int max_y = MAX(y1, y2);
+        for (int y = min_y; y < max_y; y++) {
+            for (int dx = -pad; dx < pad; dx++) {
+                for (int dy = -pad; dy < pad; dy++) {
+                    // figure out how to do this with the flex_list
+                    if (in_bounds(x1+dx, y+dy, dimension)) {
+                        bitmap[(y+dy)*dimension + (x1+dx)] = 1.0;
+                    }
+                }
+            }
+        }
+    } else {
+        for (int x = x1; x < x2; x++) {
+            int y = (int)(slope * (x - x1) + y1);
+            for (int dx = -pad; dx < pad; dx++) {
+                for (int dy = -pad; dy < pad; dy++) {
+                    if (in_bounds(x+dx, y+dy, dimension)) {
+                        bitmap[(y+dy) * dimension + (x+dx)] = 1.0;
+                    }
+                }
+            }
+        }
+    }
+    return bitmap;
+}
+
+flex_image blur_bitmap(flex_nd_vec bitmap, int ksize) {
+    std::vector<size_t> bitmap_shape = bitmap.shape();
+    flex_nd_vec blurred_bitmap(bitmap_shape, 0.0);
+    int dimension = bitmap_shape[1];
+    int pad = ksize/2;
+    for (int row = 0; row < dimension; row++) {
+        for (int col = 0; col < dimension; col++) {
+            int index = row * dimension + col;
+            if (row < pad 
+                || row >= dimension-pad 
+                || col < pad 
+                || col >= dimension-pad) {
+                blurred_bitmap[index] = (int)(255*bitmap[index]);
+                continue;
+            }
+            double sum_for_blur = 0.0;
+            for (int dr = -pad; dr <= pad; dr++) {
+                for (int dc = -pad; dc <= pad; dc++) {
+                    sum_for_blur += bitmap[(row+dr) * dimension + (col+dc)];
+                }
+            }
+            blurred_bitmap[index] = (int)(255*sum_for_blur/(ksize*ksize));
+        }
+    }
+    uint8_t image_data[dimension * dimension];
+    for (int idx=0; idx < dimension * dimension; idx++) {
+        image_data[idx] = ((uint8_t)(blurred_bitmap[idx]));
+    }
+    return flex_image((const char *)image_data,
+        dimension, dimension, 1, dimension*dimension, 1, 2);
+}
+
+flex_image rasterize(flex_list simplified_drawing) {
+    flex_image final_bitmap; // 1 x 28 x 28
+    size_t num_strokes = simplified_drawing.size();
+    int INTERMEDIATE_BITMAP_WIDTH = 256;
+    int INTERMEDIATE_BITMAP_HEIGHT = 256;
+    int FINAL_BITMAP_WIDTH = 28;
+    int FINAL_BITMAP_HEIGHT = 28;
+    float STROKE_WIDTH = 20.0;
+    int MAC_OS_STRIDE = 64;
+#ifdef __APPLE__
```

Comment:
I would really recommend extracting platform-specific code out into a separate file. Then this implementation can be tested explicitly. (For example, a Mac could explicitly test both this implementation and the portable one.)

So this could just be
```
#ifdef __APPLE__
return call_out_to_apple_specific_version(args)
#endif